assets/mdk/brittleness.mdk

```diff
@@ -0,0 +1,179 @@
+
+<!-- inline-dafny TriangleSum/A2 -->
+
+Dafny is able to prove this, but it takes over 1M RU to do so. This is partly because the postconditions from both lemmas are in scope everywhere, even though one is useful only for the base case and one is useful only for the inductive case. We can tell Dafny when to use each lemma by including a conditional statement.
+
+<!-- inline-dafny TriangleSum/A3 -->
+
+When conditioning the proof on whether `n` is zero or not, we can invoke the base case lemma only when it is, and invoke the inductive case lemma only when it is not. In addition, we use `assert by` to specialize the result of `TriangleSumRec` to the specific instantiation we need for the proof. With these two additional bits of structure, the resource use goes down to around 100K.
+
+Note that invoking the assumed lemmas about `TriangleSum` in the correct place is critical. If we use the same conditional structure but invoke both at the beginning, so they scope over the entire lemma, Dafny now uses over 1M RU again.
+
+<!-- inline-dafny TriangleSum/A4 -->
+
+# More general guidelines
+
+The examples covered so far illustrate two key principles for maintainable verification:
+
+* Break long leaps of reasoning down into smaller steps. This limits the number of reasoning steps that the prover needs to construct on its own.
+* Isolate each step so that the proof of it doesn’t need to (and, further, isn’t able to) take into account irrelevant information. This limits the number of choices that the prover has to evaluate when making each reasoning step.
+
+A nice coincidence is that programs written according to these principles tend to be more understandable to humans, as well. Reducing the work a prover needs to do when reasoning about a program has the effect of reducing the amount of thinking a human has to do to understand what a program does, and why it’s correct. Constructing small, relatively isolated components relates strongly to the concepts of [cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) and [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) used in software engineering.
+
+Those principles are rather general, however, and there are several more specific guidelines for how to apply them in practice in Dafny.
+
+* Prefer `opaque` functions, with explicit `reveal` statements where necessary.
+* Avoid `requires` and `ensures` clauses on functions, except when necessary to make them total, and prefer a collection of `lemma`s, each of which establishes a single property of the function.
```

Comment:
It would be nice to summarize as preferring extrinsic rather than intrinstic specification

Reply:
I don't want to assume that the reader knows what those terms would mean in this context, but I like the idea of mentioning them.